src/core/packages/saved-objects/base-server-internal/src/utils/pick_values_based_on_structure.test.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import { fc } from '@fast-check/jest';
+import { pickValuesBasedOnStructure, getFlattenedKeys } from './pick_values_based_on_structure';
+
+describe('getFlattenedKeys', () => {
+  test('empty object', () => {
+    expect(getFlattenedKeys({}).sort()).toEqual([]);
+  });
+
+  test('simple object', () => {
+    expect(getFlattenedKeys({ a: 1, b: 2, c: 3 }).sort()).toEqual(['["a"]', '["b"]', '["c"]']);
+  });
+
+  test('simple object 2', () => {
+    const obj = {
+      a: {},
+    };
+    expect(getFlattenedKeys(obj).sort()).toEqual(['["a"]']);
+  });
+
+  test('simple object 3', () => {
+    const obj = {
+      a: [[]],
+    };
+    expect(getFlattenedKeys(obj).sort()).toEqual(['["a"][0]']);
+  });
+
+  test('complex object 1', () => {
+    const obj = {
+      a: 1,
+      b: { c: 2 },
+      d: [{ e: 1 }, { e: 2 }],
+    };
+
+    expect(getFlattenedKeys(obj).sort()).toEqual([
+      '["a"]',
+      '["b"]["c"]',
+      '["d"][0]["e"]',
+      '["d"][1]["e"]',
+    ]);
+  });
+});
+
+describe('pickValuesBasedOnStructure', () => {
+  test('picks the values of the target', () => {
+    const a = { v1: 'a', v2: [{ c: 1, d: 1 }] };
+    const b = { v1: 'b', v2: [{ c: 2, d: 2 }, { e: 4 }], another: 'value', anArray: [1, 2, 3] };
+
+    expect(pickValuesBasedOnStructure(a, b)).toEqual({ v1: 'b', v2: [{ c: 2, d: 2 }] });
+  });
+  test('picks a subset of values from target', () => {
+    /**
+     * The `keys` `Arbitrary` represents words with possible numbers like `loremv123`
+     * that might be used in saved objects as key names.
+     */
+    const keys = fc
```

Comment:
Using `fast-check` to enable property-based testing https://fast-check.dev/docs/introduction

This is not something we do in Kibana at all, but I thought I would try it for this more critical case where we might be traversing arbitrary objects and flattening out their keys.

Do not feel super strongly about merging this test if folks feel it is too convoluted. It did help me catch a couple of bugs while developing which was nice!

Reply:
feels like a great use case for property based testing!